components/carousel/nz-carousel.component.ts

```diff
@@ -265,6 +271,95 @@ export class NzCarouselComponent implements AfterContentInit, AfterViewInit, OnD
     }
   }
 
+  private setupStream(): void {
+    const manualTransitions$ = merge(this.setupKeyboard(), this.setupSwipe(), this.setupDot()).pipe(share());
+    const timer$ = this.setupTimer(manualTransitions$);
+    merge(manualTransitions$, timer$)
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(direction => this.goTo(direction));
+    // manualTransitions$.pipe(takeUntil(this.destroy$)).subscribe((direction) => this.goTo(direction));
+  }
+
+  private setupDot(): Observable<number> {
+    return this.dotClick$;
+  }
+
+  private setupKeyboard(): Observable<number> {
+    const carouselElement = this.slickListEl;
+    const leftArrow$ = fromEvent(carouselElement, 'keydown').pipe(
+      filter((event: Event) => (event as KeyboardEvent).keyCode === LEFT_ARROW),
+      tap(e => e.preventDefault()),
+      map(() => this.activeIndex - 1)
+    );
+    const rightArrow$ = fromEvent(carouselElement, 'keydown').pipe(
+      filter((event: Event) => (event as KeyboardEvent).keyCode === RIGHT_ARROW),
+      tap(e => e.preventDefault()),
+      map(() => this.activeIndex + 1)
+    );
+    return merge(leftArrow$, rightArrow$);
+  }
+
+  private setupSwipe(): Observable<number> {
+    const carouselElement = this.slickListEl;
+    const start$ = merge(fromEvent(carouselElement, 'mousedown'), fromEvent(carouselElement, 'touchdown')).pipe(
+      tap(() => (this.gestureRect = this.slickListEl.getBoundingClientRect()))
+    );
+    const move$ = (startEvent: MouseEvent | TouchEvent) =>
+      this.nzDragService.requestDraggingSequence(startEvent).pipe(
+        observeOn(animationFrameScheduler), // performance
+        tap(delta => {
+          this.strategy.dragging(delta);
+        }), // do animation here
+        takeLast(1)
+      );
+    return start$.pipe(
+      switchMap(startEvent => move$(startEvent as TouchEvent | MouseEvent)),
+      map(delta => {
+        const xDelta = delta.x;
+        const direction = Math.abs(xDelta) > this.gestureRect!.width / 3 ? (xDelta > 0 ? -1 : 1) : 0;
+        return this.activeIndex + direction;
+      }),
+      tap(() => (this.gestureRect = null))
+    );
+  }
+
+  private setupTimer(manualTransitions$: Observable<number>): Observable<number> {
+    return this.autoPlayChange$.pipe(
+      startWith(true),
+      map(() => this.nzAutoPlaySpeed),
+      switchMap(speed =>
+        interval(speed)
+          .pipe(
+            takeUntil(manualTransitions$),
+            repeatWhen(self => self),
+            filter(() => this.nzAutoPlay),
+            map(() => this.activeIndex + 1)
+          )
+          .pipe(finalize(() => console.log('I am removed')))
```

Comment:
Will come back to this later. No idea why Jasmine complains that there's still a periodic timer in the queue. They should all be removed.